include/zephyr/devicetree/mem.h

```diff
@@ -0,0 +1,111 @@
+/**
+ * @file
+ * @brief Memory Devicetree macro public API header file.
+ */
+
+/*
+ * Copyright (c) 2020, Linaro Ltd.
+ * Copyright (c) 2020 Nordic Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DEVICETREE_MEM_H_
+#define ZEPHYR_INCLUDE_DEVICETREE_MEM_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include <string.h>
+
+/**
+ * @defgroup devicetree-memory Devicetree memory API
+ * @ingroup devicetree
+ * @{
+ */
+
+/* Inidicates that a ram region is nocache. */
+/* TODO Obtained from zephyr,memory-region.yaml. This should be defined in
+ * just one place.
+ */
+#define MEM_NOCACHE_ATTR "RAM_NOCACHE"
+
+/**
+ * @brief Container for MMIO SRAM information specified in devicetree
+ *
+ * This type contains pointers to the begin and end of the memory region and
+ * a string that specifies such region's properties.
+ *
+ */
+struct dt_mem_region {
+	const uint8_t *start;
+	const uint8_t *end;
+	const char *region_mpu;
+};
+
+/**
+ * @brief Can be used to initialize a @see{struct dt_mem_region}.
+ *
+ * @param node_id Node ID of the memory region declared in device tree from
+ * which the information will be obtained.
+ *
+ * @example
+ *
+ *		struct dt_mem_region mem_r = {
+ *			DT_GET_MEM_REGION(DT_NODELABEL(<node_label>))
+ *		};
+ *
+ */
+#define DT_GET_MEM_REGION(node_id)                                              \
+	{                                                                           \
+		.start = (const uint8_t *)DT_REG_ADDR(node_id),                         \
+		.end =                                                                  \
+			(const uint8_t *)(DT_REG_ADDR(node_id) + DT_REG_SIZE(node_id)),     \
+		.region_mpu = COND_CODE_1(                                              \
+			DT_NODE_HAS_PROP(node_id, zephyr_memory_region_mpu),                \
+			(DT_PROP(node_id, zephyr_memory_region_mpu)),                       \
+			(NULL)                                                              \
+		)                                                                       \
+	},
+
+/**
+ * Return true if buf is in a nocache memory region
+ * among mem_regions.
+ *
+ * @param buf Buffer to be checked
+ * @param size Size of the buffer above
+ * @param mem_regions Memory regions agaisnt which buf will be checked.
+ * @param mem_regions_cnt Number of memory regions
+ */
+static inline bool dt_buf_in_nocache(
+			const uint8_t *buf,
+			size_t size,
+			const struct dt_mem_region mem_regions[],
```

Comment:
Is there a way we could avoid this parameter by walking DT for all memory regions perhaps and coalescing them into an iterable section of dt_mem_region structs?